x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtSignatureValidator.java

```diff
@@ -0,0 +1,316 @@
+            final boolean isConfiguredJwkSetHmac = Strings.hasText(hmacJwkSetContents);
+            final boolean isConfiguredJwkOidcHmac = Strings.hasText(hmacKeyContents);
+            validateJwkSettings(realmConfig, isConfiguredJwkSetPkc, isConfiguredJwkSetHmac, isConfiguredJwkOidcHmac);
+
+            final List<JWK> jwksHmac;
+            if (isConfiguredJwkSetHmac) {
+                jwksHmac = JwkValidateUtil.loadJwksFromJwkSetString(
+                    RealmSettings.getFullSettingKey(realmConfig, JwtRealmSettings.HMAC_JWKSET),
+                    hmacJwkSetContents.toString()
+                );
+            } else if (isConfiguredJwkOidcHmac) {
+                final OctetSequenceKey hmacKey = JwkValidateUtil.loadHmacJwkFromJwkString(
+                    RealmSettings.getFullSettingKey(realmConfig, JwtRealmSettings.HMAC_KEY),
+                    hmacKeyContents
+                );
+                assert hmacKey != null : "Null HMAC key should not happen here";
+                jwksHmac = List.of(hmacKey);
+            } else {
+                jwksHmac = null;
+            }
+
+            if (jwksHmac != null) {
+                final JwkSetLoader.JwksAlgs jwksAlgs = JwkValidateUtil.filterJwksAndAlgorithms(jwksHmac, allowedJwksAlgsHmac);
+                logger.info("Usable HMAC: JWKs [{}]. Algorithms [{}].", jwksAlgs.jwks().size(), String.join(",", jwksAlgs.algs()));
+                // Filter JWK(s) vs signature algorithms. Only keep JWKs with a matching alg. Only keep algs with a matching JWK.
+                this.hmacJwtSignatureValidator = new HmacJwtSignatureValidator(jwksAlgs);
+            } else {
+                this.hmacJwtSignatureValidator = null;
+            }
+
+            if (isConfiguredJwkSetPkc) {
+                this.pkcJwtSignatureValidator = new PkcJwtSignatureValidator(new JwkSetLoader(realmConfig, sslService), reloadNotifier);
+            } else {
+                this.pkcJwtSignatureValidator = null;
+            }
+
+            if (false == hasUsableJwksAlgs()) {
+                logger.warn(
+                    "No available JWK and algorithm for HMAC or PKC. JWT realm authentication expected to fail until this is fixed."
+                );
+            }
+        }
+
+        @Override
+        public void validate(String tokenPrincipal, SignedJWT jwt, ActionListener<Void> listener) {
+            final String algorithm = jwt.getHeader().getAlgorithm().getName();
+            if (allowedJwksAlgsHmac.contains(algorithm)) {
+                if (hmacJwtSignatureValidator != null) {
+                    hmacJwtSignatureValidator.validate(tokenPrincipal, jwt, listener);
+                } else {
+                    listener.onFailure(
+                        new ElasticsearchSecurityException(
+                            "algorithm [%s] is a HMAC signing algorithm, but HMAC JWK is not configured",
+                            RestStatus.BAD_REQUEST,
+                            algorithm
```

Comment:
Should these error messages mention the relevant settings to configure?

Reply:
Added here as well as the similar error message below this.